Mathlib/GroupTheory/GroupAction/Period.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2024 Emilie Burgun. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Emilie Burgun
+-/
+
+import Mathlib.Dynamics.PeriodicPts
+import Mathlib.GroupTheory.Exponent
+import Mathlib.GroupTheory.GroupAction.Basic
+
+/-!
+# Period of a group action
+
+This module defines some helpful lemmas around [`MulAction.period`] and [`AddAction.period`].
+The period of a point `a` by a group element `g` is the smallest `m` such that `g ^ m • a = a`
+(resp. `(m • g) +ᵥ a = a`) for a given `g : G` and `a : α`.
+
+If such an `m` does not exist,
+then by convention `MulAction.period` and `AddAction.period` return 0.
+-/
+
+namespace MulAction
+
+universe u v
+variable {α : Type v}
+variable {G : Type u} [Group G] [MulAction G α]
+variable {M : Type u} [Monoid M] [MulAction M α]
+
+/-- If the action is periodic, then a lower bound for its period can be computed. -/
+@[to_additive "If the action is periodic, then a lower bound for its period can be computed."]
+theorem le_period {m : M} {a : α} {n : ℕ} (period_pos : 0 < period m a)
+    (moved : ∀ k, 0 < k → k < n → m ^ k • a ≠ a) : n ≤ period m a :=
+  le_of_not_gt fun period_lt_n =>
+    moved _ period_pos period_lt_n <| pow_period_smul m a
+
+/-- If for some `n`, `m ^ n • a = a`, then `period m a ≤ n`. -/
+@[to_additive "If for some `n`, `(n • m) +ᵥ a = a`, then `period m a ≤ n`."]
+theorem period_le_of_fixed {m : M} {a : α} {n : ℕ} (n_pos : 0 < n) (fixed : m ^ n • a = a) :
+    period m a ≤ n :=
+  (isPeriodicPt_smul_iff.mpr fixed).minimalPeriod_le n_pos
+
+/-- If for some `n`, `m ^ n • a = a`, then `0 < period m a`. -/
+@[to_additive "If for some `n`, `(n • m) +ᵥ a = a`, then `0 < period m a`."]
+theorem period_pos_of_fixed {m : M} {a : α} {n : ℕ} (n_pos : 0 < n) (fixed : m ^ n • a = a) :
+    0 < period m a :=
+  (isPeriodicPt_smul_iff.mpr fixed).minimalPeriod_pos n_pos
+
+@[to_additive]
+theorem period_eq_one_iff {m : M} {a : α} : period m a = 1 ↔ m • a = a := ⟨
+  fun eq_one => pow_one m ▸ eq_one ▸ pow_period_smul m a,
+  fun fixed => le_antisymm
+    (period_le_of_fixed one_pos (by simpa))
+    (period_pos_of_fixed one_pos (by simpa))⟩
```

Comment:
For some reason the style guide says brackets belong on the next line, even if it makes indentation look slightly ugly.
```suggestion
theorem period_eq_one_iff {m : M} {a : α} : period m a = 1 ↔ m • a = a :=
  ⟨fun eq_one => pow_one m ▸ eq_one ▸ pow_period_smul m a,
   fun fixed => le_antisymm
    (period_le_of_fixed one_pos (by simpa))
    (period_pos_of_fixed one_pos (by simpa))⟩
```